modules/tensor_mechanics/src/materials/ComputeLinearElasticPFFractureStress.C

```diff
@@ -8,112 +8,247 @@
+
+  // Compute eigenvectors and eigenvalues of mechanical strain and projection tensor
+  RankTwoTensor eigvec;
+  std::vector<Real> eigval(LIBMESH_DIM);
+  RankFourTensor Ppos =
+      _mechanical_strain[_qp].positiveProjectionEigenDecomposition(eigval, eigvec);
+  RankFourTensor I4sym(RankFourTensor::initIdentitySymmetricFour);
+  RankFourTensor Pneg = I4sym - Ppos;
 
-  // Zero out values when c > 1
-  Real cfactor = 1.0;
-  if (c > 1.0)
-    cfactor = 0.0;
+  // Calculate tensors of outerproduct of eigen vectors
+  std::vector<RankTwoTensor> etens(LIBMESH_DIM);
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+      for (unsigned int k = 0; k < LIBMESH_DIM; ++k)
+        etens[i](j, k) = eigvec(j, i) * eigvec(k, i);
+
+  // Separate out positive and negative eigen values
+  std::vector<Real> epos(LIBMESH_DIM), eneg(LIBMESH_DIM);
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+  {
+    epos[i] = (std::abs(eigval[i]) + eigval[i]) / 2.0;
+    eneg[i] = (std::abs(eigval[i]) - eigval[i]) / 2.0;
+  }
+
+  // Seprate positive and negative sums of all eigenvalues
+  Real etr = 0.0;
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    etr += eigval[i];
+
+  const Real etrpos = (std::abs(etr) + etr) / 2.0;
+  const Real etrneg = (std::abs(etr) - etr) / 2.0;
+
+  // Calculate the tensile (postive) and compressive (negative) parts of stress
+  RankTwoTensor stress0pos, stress0neg;
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+  {
+    stress0pos += etens[i] * (lambda * etrpos + 2.0 * mu * epos[i]);
+    stress0neg += -etens[i] * (lambda * etrneg + 2.0 * mu * eneg[i]);
+  }
+
+  // sum squares of epos and eneg
+  Real pval(0.0), nval(0.0);
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+  {
+    pval += epos[i] * epos[i];
+    nval += eneg[i] * eneg[i];
+  }
+
+  _stress[_qp] = stress0pos * _D[_qp] + stress0neg;
+
+  // Energy with positive principal strains
+  F_pos = lambda * etrpos * etrpos / 2.0 + mu * pval;
+  F_neg = -lambda * etrneg * etrneg / 2.0 + mu * nval;
 
+  // 2nd derivative wrt c and strain = 0.0 if we used the previous step's history varible
+  if (_use_current_hist)
+    _d2Fdcdstrain[_qp] = stress0pos * _dDdc[_qp];
+
+  // Used in StressDivergencePFFracTensors off-diagonal Jacobian
+  _dstress_dc[_qp] = stress0pos * _dDdc[_qp];
+
+  // Calculate dstress/dstrain
+  // dstress/dstrain = dstress_pos/dstrain_pos * dstrain_pos/dstrain
+  //                 + dstress_neg/dstrain_neg * dstrain_neg/dstrain
+  // dstrain_pos/dstrain = proj_pos (fourth order projection tensor)
+  // dstrain_neg/dstrain = proj_neg (fourth order projection tensor)
+  // Note that proj_pos + proj_neg = symmetric identify fourth order tensor
+
+  // d trace(A) / dA
+  RankTwoTensor I(RankTwoTensor::initIdentity);
+  RankFourTensor dtraceAdA = I.outerProduct(I);
+
+  _Jacobian_mult[_qp] = _D[_qp] * (lambda * dtraceAdA * MathUtils::heavyside(etr) + 2 * mu * Ppos) +
```

Comment:
We can also save computation time here.
Currently this line does twice RankFourTensor multiplications.
However, if we write Jacobian = [Identity - (1 - degradation) * Ppos] * elasticity_tensor,
then there will be only one RankFourTensor multiplication.

Reply:
good!